### PR TITLE
Add route controller for better composability. (#565)

### PR DESCRIPTION
**Pull request by connor-ricks**
### Description

Adds `RouteController` to improve the composability of routes when using `HummingbirdRouter`

Conforming to `RouteController` allows you to create composable routes in a more `SwiftUI` result builder style.

```swift
struct ChildAController: RouteController {
    typealias Context = BasicRouterRequestContext
    var body: some RouterMiddleware<Context> {
        RouteGroup("/child_a") {
            Get { _,_ in "child_a" }
        }
    }
}

struct ChildBController: RouteController {
    typealias Context = BasicRouterRequestContext
    var body: some RouterMiddleware<Context> {
        RouteGroup("/child_b") {
            Get { _,_ in "child_b" }
        }
    }
}

struct ParentController: RouteController {
    typealias Context = BasicRouterRequestContext
    var body: some RouterMiddleware<Context> {
        RouteGroup("/parent") {
            Get { _,_ in return "parent" }
            ChildAController()
            ChildBController()
            Get("child_c") { _,_ in "child_c" }
        }
    }
}
```

### CODE REVIEW

File: Sources/HummingbirdRouter/RouteController.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+///             RouteGroup("/users") {
+///                 ...
+///             }
+///         }
+///     }
+///
+/// Assemble the controller's body by combining one or more components together to create
+/// a route. By nesting controllers within each other, you can compose large and complex
+/// routes from smaller more mangeable components.
+public protocol RouteController where Context == Body.Context {
```
Comment (Joannis) [inline Sources/HummingbirdRouter/RouteController.swift:38]: A primary associated type would be nice, so you can filter for `RouteController<MyContextType>`

```suggestion
public protocol RouteController<Context> where Context == Body.Context {
```

**Review by Joannis — COMMENTED** [review]
Looks like an excellent addition, thank you! One minor nit from my end - and a review by @adam-fowler who's on vacation right now.

File: Sources/HummingbirdRouter/RouteController.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    public static func buildBlock<C0: RouteController>(_ c0: C0) -> C0.Body {
+        return c0.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController>(first: C0) -> C0.Body {
+        first.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController, C1: RouteController>(
+        accumulated c0: C0,
```
Comment (adam-fowler) [inline Sources/HummingbirdRouter/RouteController.swift:64]: The accumulated middleware should only ever be a MiddlewareProtocol so I don't think this function is necessary.

File: Sources/HummingbirdRouter/RouteController.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        next c1: C1
+    ) -> _Middleware2<C0.Body, C1.Body> where C0.Body.Input == C1.Body.Input, C0.Body.Output == C1.Body.Output, C0.Body.Context == C1.Body.Context {
+        _Middleware2(c0.body, c1.body)
+    }
+
+    public static func buildPartialBlock<C0: RouteController, M0: MiddlewareProtocol>(
+        accumulated c0: C0,
+        next m0: M0
+    ) -> _Middleware2<C0.Body, M0> where C0.Body.Input == M0.Input, C0.Body.Output == M0.Output, C0.Body.Context == M0.Context {
```
Comment (adam-fowler) [inline Sources/HummingbirdRouter/RouteController.swift:73]: The accumulated middleware should only ever be a MiddlewareProtocol so I don't think this function is necessary.

File: Sources/HummingbirdRouter/RouteController.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+extension RouteBuilder {
+    public static func buildExpression<C0: RouteController>(_ c0: C0) -> C0.Body where C0.Body.Input == Request, C0.Body.Output == Response, C0.Body.Context == Context {
+        c0.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController>(first: C0) -> C0.Body {
+        first.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController, C1: RouteController>(
```
Comment (adam-fowler) [inline Sources/HummingbirdRouter/RouteController.swift:96]: The accumulated middleware should be a MiddlewareProtocol.

Comment (adam-fowler) [reply Sources/HummingbirdRouter/RouteController.swift:96]: Although I'm not sure we need to support this for the RouteBuilder result builder. This is only ever used in the situation where the endpoint is already selected eg

```swift
Get("test") {
    MyMiddleware()
    Handle { request, context in
        ...
    }
```

File: Sources/HummingbirdRouter/RouteController.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// MARK: - MiddlewareFixedTypeBuilder + RouteController
+
+/// Middleware stack result builder
+///
+/// Generates a middleware stack from the elements inside the result builder. The input,
+/// context and output types passed through the middleware stack are fixed and cannot be changed.
+extension MiddlewareFixedTypeBuilder {
+    public static func buildExpression<C0: RouteController>(_ c0: C0) -> C0.Body where C0.Body.Input == Input, C0.Body.Output == Output, C0.Body.Context == Context {
```
Comment (adam-fowler) [inline Sources/HummingbirdRouter/RouteController.swift:51]: I've written all the other MiddlewareFixedTypeBuilder functions using MiddlewareProtocol. It would be great if you could continue that.

**Review by adam-fowler — CHANGES_REQUESTED** [review]

File: Sources/HummingbirdRouter/RouteController.swift (hunk 5 above)
Comment (connor-ricks) [reply Sources/HummingbirdRouter/RouteController.swift:51]: I'm not sure I follow the ask here. Given that we decided not to have the `RouteController` conform to `MiddlewareProtocol`, I believe I have to conform the build expression's generics to `RouteController` in order to compose them together.

I might be misunderstanding the ask though.

Comment (connor-ricks) [reply Sources/HummingbirdRouter/RouteController.swift:51]: I'm not sure I follow the ask here. Given that `RouteController` itself doesn't conform to `MiddlewareProtocol` I think I need to constrain the generic to `RouteController`. I might be misunderstanding the ask though.

Comment (connor-ricks) [reply Sources/HummingbirdRouter/RouteController.swift:51]: I updated the `RouteController` to have a body that is a `MiddlewareProtocol` and I also changed it to use `@MiddlewareFixedTypeBuilder` instead. Maybe that's what you were referring to?

**Comment by connor-ricks** [comment]
There's definitely some ergonomic improvements that could be made here.

Creating a high order controller with generic content gets a bit wordy.

It's also not super intuitive. I wonder what improvements we can make to simplify the concept.

I'll be away from my comp for a few days, so won't be able to contribute much beyond discussion, but feel free to amend to this commit.

```swift
        struct FooContainerController<Content: RouterMiddleware>: RouteController where Content.Context: RouterRequestContext {
            public typealias Context = Content.Context

            @RouteBuilder<Context> var content: () -> Content

            var body: some RouterMiddleware<Context> {
                content()
            }
        }

        struct BarContainerController<Content: MiddlewareProtocol>: RouteController where Content.Context: RouterRequestContext {
            public typealias Context = Content.Context

            @MiddlewareFixedTypeBuilder<Content.Input, Content.Output, Content.Context> var content: () -> Content

            var body: some MiddlewareProtocol<Content.Input, Content.Output, Content.Context> {
                content()
            }
        }
```
